[pt2e][quant] Add weight only quantization for xnnpack

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118454

Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[pt2e][quant] Add weight only quantization for xnnpack

Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

digantdesai commented on test/quantization/pt2e/test_xnnpack_quantizer.py:
```diff
@@ -582,6 +582,63 @@ def test_dynamic_linear_int4_weight(self):
+        quantization_config = get_symmetric_quantization_config(
+            is_per_channel=True,
+            is_weight_only=True,
+            weight_qmin=0,
+            weight_qmax=15,
```
```suggestion
            weight_qmin=-8,
            weight_qmax=7,
```

jerryzh168 replied:
I thought xnnpack is using 0, 15? that's why they have zero_point = 8 right?

digantdesai replied:
This feels like better UX with our symmetric quant theme for xnnpack. We can shift the weights to zp=8 and make then uint8 during the lowering.

digantdesai reviewed (approved):
Looks good. Thanks. I can wire this up at least for fp32 now. It will take more work to do fp16.

jerryzh168 committed:
Update on "[pt2e][quant] Add weight only quantization for xnnpack"


Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[pt2e][quant] Add weight only quantization for xnnpack"


Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[pt2e][quant] Add weight only quantization for xnnpack"


Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[pt2e][quant] Add weight only quantization for xnnpack"


Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot rebase

jerryzh168 committed:
Update on "[pt2e][quant] Add weight only quantization for xnnpack"


Summary:
with fp32 activations

Test Plan:
python test/test_quantization.py -k test_weight_only_linear_int4_weight

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot rebase
